tools/tiller-hack/tiller-hack.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package main provides a tool tiller-hack, this tools allows exclude further
+// tracking of CRDs objects by helm's tiller after they were deployed by means of
+// helm's charts.
+//
+// ./tiller-hack --kubeconfig={location of kube config file} \
+//               --configmap={the name of tiller's config map for istio deployment}
+//               --namespace={the namespace of configmap and tiller,
+//                            by default it is kube-system}
+//               --restart-tiller={true/false, false is default, true will
+//                                 force the tool to restart tiller}
+//
+package main
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/base64"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/retry"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/golang/protobuf/proto"
+	"k8s.io/helm/pkg/proto/hapi/chart"
+	rspb "k8s.io/helm/pkg/proto/hapi/release"
+)
+
+var (
+	kubeconfig = flag.String("kubeconfig", "",
+		"Absolute path to the kubeconfig file, must be specified.")
+	configMapName = flag.String("configmap", "",
+		"Name of the configmap to patch, must be specified.")
+	configMapNamespace = flag.String("namespace", "kube-system",
+		"Namespace of the configmap, defaults to \"kube-system\"")
+	restartTillerFlag = flag.Bool("restart-tiller", false,
+		"Restart tiller after patching the configmap")
+
+	// In older versions of helm, the compression of release information has not been used.
+	// This header identify whether the following info is compressed or not.
+	magicGzip = []byte{0x1f, 0x8b, 0x08}
+)
+
+// BuildClient returns kubernetes clientset
+func BuildClient(kubeconfig string) (*kubernetes.Clientset, error) {
+	var config *rest.Config
+	var err error
+
+	kubeconfigEnv := os.Getenv("KUBECONFIG")
+
+	if kubeconfigEnv != "" {
+		kubeconfig = kubeconfigEnv
+	}
+
+	if kubeconfig != "" {
+		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+	} else {
+		config, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		return nil, err
+	}
+	k8s, err := kubernetes.NewForConfig(config)
+	if err != nil {
```

Comment:
reading this code vs the usage...  if --kubeconfig="", this option will read from the environment, ten the in cluster secretg.  If not set to "" a file will be used.  Why force the specification of this value?

Reply:
because in most of the cases this code will be run outside of cluster and people normally should be given a choice which config cluster to use. Example I have 5 clusters and by selecting corresponding cluster config this tool talks to corresponding cluster.